### PR TITLE
ranger: fix the range calculation when we meet collation

### LINKED ISSUE #14695: expression: update operator `IN` to support collations.
## Feature Request

**Describe the feature you'd like:**

In order to provide full collation support in TiDB, the implementation of operator `IN` should be updated to be collation-aware. For example, the following result is not true in TiDB for now:

```
mysql> SELECT 'A' IN ('À');                                                                     
+--------------+
| 'A' IN ('À') |
+--------------+
| 1            |
+--------------+
1 row in set
```

**Teachability, Documentation, Adoption, Migration Strategy:**

Please see TiDB RFC #14574 for more details on collations support in TiDB.

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve? 

We should check the collation information when detaching the cond instead of when building ranges.
We missed to deal with the `IN` function.

This PR closes #14695.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

**Comment by winoros** [comment]
./run-all-tests

File: util/ranger/ranger_test.go [hunk 1]
```diff
@@ -584,8 +598,31 @@ func (s *testRangerSuite) TestIndexRange(c *C) {
...
+		{
+			indexPos:    4,
+			exprStr:     "f = 'a' and f = 'B' collate utf8mb4_bin",
+			accessConds: "[eq(test.t.f, a)]",
+			filterConds: "[eq(test.t.f, B)]",
+			resultStr:   "[[\"a\",\"a\"]]",
+		},
 	}
 
+	collate.SetNewCollationEnabledForTest(true)
```
Comment (wjhuang2016) [inline util/ranger/ranger_test.go]: The testSuite must be serial.

File: util/ranger/checker.go [hunk 2]
```diff
@@ -39,17 +40,24 @@ func (c *conditionChecker) check(condition expression.Expression) bool {
...
 			if c.checkColumn(scalar.GetArgs()[1]) {
+				if scalar.GetArgs()[1].GetType().EvalType() == types.ETString && !collate.CompatibleCollate(scalar.GetArgs()[1].GetType().Collate, collation) {
+					return false
+				}
 				return scalar.FuncName.L != ast.NE || c.length == types.UnspecifiedLength
 			}
 		}
 		if _, ok := scalar.GetArgs()[1].(*expression.Constant); ok {
 			if c.checkColumn(scalar.GetArgs()[0]) {
+				if scalar.GetArgs()[1].GetType().EvalType() == types.ETString && !collate.CompatibleCollate(scalar.GetArgs()[0].GetType().Collate, collation) {
```
Comment (bb7133) [inline util/ranger/checker.go]:
```suggestion
				if scalar.GetArgs()[0].GetType().EvalType() == types.ETString && !collate.CompatibleCollate(scalar.GetArgs()[0].GetType().Collate, collation) {
```
Should it be [0] here?

File: util/ranger/checker.go [hunk 3]
```diff
@@ -39,17 +40,24 @@ func (c *conditionChecker) check(condition expression.Expression) bool {
...
 
 func (c *conditionChecker) checkScalarFunction(scalar *expression.ScalarFunction) bool {
+	_, collation, _ := scalar.CharsetAndCollation(scalar.GetCtx())
 	switch scalar.FuncName.L {
 	case ast.LogicOr, ast.LogicAnd:
 		return c.check(scalar.GetArgs()[0]) && c.check(scalar.GetArgs()[1])
 	case ast.EQ, ast.NE, ast.GE, ast.GT, ast.LE, ast.LT:
 		if _, ok := scalar.GetArgs()[0].(*expression.Constant); ok {
 			if c.checkColumn(scalar.GetArgs()[1]) {
+				if scalar.GetArgs()[1].GetType().EvalType() == types.ETString && !collate.CompatibleCollate(scalar.GetArgs()[1].GetType().Collate, collation) {
```
Comment (bb7133) [inline util/ranger/checker.go]: Please add some comments here.

**Pull request merged**